### PR TITLE
Provide target allocation id as part of start recovery request

### LINKED ISSUE #24167: CI Failure: org.elasticsearch.recovery.RelocationIT testIndexAndRelocateConcurrently
Unable to reproduce locally.

CI Failure:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos/852/consoleFull

Reproduce With:
gradle :core:integTest -Dtests.seed=4856A7A6357BA799 -Dtests.class=org.elasticsearch.recovery.RelocationIT -Dtests.method="testIndexAndRelocateConcurrently" -Dtests.security.manager=true -Dtests.locale=sk -Dtests.timezone=America/Argentina/Rio_Gallegos

### PR DESCRIPTION
**Pull request by ywelsch**
This makes it possible for the recovery source to verify that it is talking to the shard it thinks it is talking to.

Closes #24167

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java [hunk 1]
```diff
@@ -113,7 +113,22 @@ private RecoveryResponse recover(final StartRecoveryRequest request) throws IOEx
             throw new DelayRecoveryException("source node has the state of the target shard to be [" + targetShardRouting.state() + "], expecting to be [initializing]");
         }
 
-        RecoverySourceHandler handler = ongoingRecoveries.addNewRecovery(request, targetShardRouting.allocationId().getId(), shard);
+        // the following IF condition won't go into master, it only goes into 5.4 / 5.x branch to allow for interoperability with < 5.4
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java]: At which point will you also put back the `final` modifier on the `request` method parameter?

**Review by jasontedor — APPROVED** [review]
The code change LGTM, but all the work is in the analysis in #24167 which is great. Thanks for digging so deeply into this one.

File: core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java [hunk 2]
```diff
@@ -108,6 +114,7 @@ public void readFrom(StreamInput in) throws IOException {
         super.readFrom(in);
         recoveryId = in.readLong();
         shardId = ShardId.readShardId(in);
+        targetAllocationId = in.readString();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: this needs some version conditioned reads / writes?!

**Review by s1monw — CHANGES_REQUESTED** [review]
I left a question

File: core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java (hunk 2 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: yeah, i just removed this in a follow-up commit for master (ebad5a8). The first commit will be the one that goes into 5.4 and 5.x

File: core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java]: I've pushed ebad5a8 which will be the version that goes into master

**Review by s1monw — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: alright!

**Comment by ywelsch** [comment]
Thanks @jasontedor @s1monw

**Pull request merged**